website/registrations/utils.py

```diff
@@ -0,0 +1,376 @@
+        # self.raw_csv = bulk_upload_csv.read().decode('utf-8')
+        self.reader = csv.DictReader(io.StringIO(self.raw_csv))
+        self.headers = [header.lower() for header in self.reader.fieldnames]
+        schema_id_row = next(self.reader)
+        self.schema_id = schema_id_row[self.reader.fieldnames[0]]
+        self.provider_id = provider_id
+        if self.schema_id:
+            try:
+                self.registration_provider = RegistrationProvider.load(self.provider_id)
+                self.registration_schema = self.registration_provider.schemas.get(_id=self.schema_id)
+            except RegistrationSchema.DoesNotExist:
+                raise NotFound(detail='Schema with id "{}" was not found'.format(self.schema_id))
+            except RegistrationProvider.DoesNotExist:
+                raise NotFound(detail='Registration provider with id "{}" was not found').format(self.provider_id)
+        self.schema_questions = BulkRegistrationUpload.get_schema_questions_validations(self.registration_schema)
+        self.validations = {**self.schema_questions, **METADATA_FIELDS}
+        self.errors = []
+        self.rows = [Row(row,
+                         self.validations,
+                         functools.partial(self.log_error, row_index=self.reader.line_num))
+                     for row in self.reader]
+
+    @classmethod
+    def get_schema_questions_validations(cls, registration_schema):
+        schema_questions = {}
+        get_question_validations = lambda question: {'required': question.get('required', False),
+                                   'format': question.get('format', ''),
+                                   'type': question.get('type', 'string'),
+                                   'options': question.get('options', [])}
+        for page in registration_schema.schema['pages']:
+            for question in page['questions']:
+                schema_questions[question['qid']] = get_question_validations(question)
+                if 'properties' in question:
+                    for nested_question in question['properties']:
+                        qid = '{}.{}'.format(question['qid'], nested_question['id'])
+                        schema_questions[qid] = get_question_validations(nested_question)
+        return schema_questions
+
+    def log_error(self, **kwargs):
+        self.errors.append({
+            'header': kwargs['header'],
+            'column_index': get_excel_column_name(kwargs['column_index']),
+            'row_index': kwargs['row_index'],
+            'missing': kwargs.get('missing', False),
+            'invalid': kwargs.get('invalid', False),
+            'external_id': kwargs.get('external_id', ''),
+            'type': kwargs.get('type', '')
+        })
+
+    def validate_csv_header_list(self):
+        expected_headers = self.validations.keys()
+        actual_headers = self.headers
+        diff = set(expected_headers) - set(actual_headers)
```

Comment:
Is `set` diff case-sensitive? If so, we need to update the metadata fields to be exactly the same as what we have in the templates.

Reply:
It seems when we initialize the class, the headers have been converted to lower case. So even if `set` diffs are case-sensitive, it shouldn't matter here.